### ISSUE TITLE
Relinker not updating FormIDs

### ISSUE THREAD

**Issue by ArcaneTourist** [matortheeternal/zedit#162]
I have several merges.  The relinker correctly fixes the esp names.   However, none of the changed FormIDs are fixed up.   The merge process does list them in the map.json.   However, there are no messages at all from fixGetFormCalls in the relinker log.   Is there anything I can do to help narrow this down short of building zedit myself and adding debugging?

**Comment by matortheeternal** [comment]
Hi Michael, thanks for the report.

Are you sure that the formIDs changed?  zMerge selectively renumbers formIDs when there's a conflict so it's possible that the formIDs referenced in the papyrus scripts have not changed.

**Comment by ArcaneTourist** [comment]
Yes, they're listed in the map.json files and I initially discovered the issue due to Papyrus errors in a script that couldn't find the formids it expected.

FWIW, my mod list is Lexy's LOTD with some additions, so I have a lot of merges.

I tried to build zedit so that I could try adding debugging to relinker.js.   I got as far as figuring out that I needed nodejs 8.x and not newer versions.  However, "npm start" does nothing and "npm run release" produces stuff in a win-unpacked folder that fails immediately with a popup error "critical error loading XEditLib.dll: GetGlobal failed  at helpers.Fail ... app.js:1041".   I'm a programmer, but all my experience is with Unix and Linux and I don't really know anything about Windows frameworks.

I can't imagine that this is affecting everyone.  Nor can I see why my configuration should be any different than anyone elses.   FWIW, I'm running on Win 8.1 which seems to be a bit rare.   I'd hope that's not the issue though.

**Comment by matortheeternal** [comment]
I'll look into it for sure.  In regards to debugging, you can just open the inspector with Ctrl+Shift+I, search the code for the appropriate service, and then place breakpoints.  You can search for "relinker" and find the declaration for `ngapp.service('relinker'` then just drop a breakpoint in `getScriptsToRelink` or `fixGetFormCalls`.

**Comment by Targuinia** [comment]
So I just ran into this problem too, so I downloaded pex-parser and investigated for a bit. 

It seems game.GetFormFromFile uses callstatic (opcode 25, 0x19) rather than callmethod (opcode 23, 0x17), so fixGetFormCalls will of course always return at `if (n.op !== CALLMETHOD.code) return;`.

Looking for callstatic instead (and changing the indices for `n.arguments`) will properly change the formids.

The changed code:
```
let fixGetFormCalls = function(script, fidMap) {
    let functions = pexService.getFunctions(script);
    functions.forEach(fn => {
        fn.instructions.forEach(n => {
            if (n.op !== 25) return;
            let methodName = resolveString(script, n.arguments[1]);
            if (methodName !== 'GetFormFromFile') return;
            let filename = resolveString(script, n.arguments[5]);
            log(`Found GetFormFromFile call targetting ${filename}`, true);
            if (!fidMap.hasOwnProperty(filename)) return;
            let oldFormId = xelib.Hex(n.arguments[4].data, 6),
                newFormId = fidMap[filename][oldFormId];
            if (!newFormId) return log(`Form ID ${oldFormId} not renumbered`);
            log(`Changing Form ID from ${oldFormId} to ${newFormId}`);
            n.arguments[4].data = parseInt(newFormId, 16);
        });
    });
};
```

I wanted to test with a proper build and make a merge request, but I am getting the same error when starting the app as above.

Maybe there's some edge cases I didn't think of, since I am not super familiar with Papyrus, but after decompiling scripts and some testing in game, it seems to work properly like this.

**Comment by pijaczkar8** [comment]
Has this issue been fixed yet?  I'm still having the issue with formIDs not being updated to match renumbered IDs as per the merge maps.

**Comment by ArcaneTourist** [comment]
I see a commit seven days ago saying "Fixed relinker not handling renumbered formids properly."

I haven't tested, but I'm guessing this bug is fixed?  If so, hoooray and thanks!

Can you please put up a new release?

**Comment by matortheeternal** [comment]
Issue is fixed, new release on its way.
